File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/ListUtils.java
package com.garpr.android.misc;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.garpr.android.R;
import com.garpr.android.models.AbsPlayer;
import com.garpr.android.models.AbsTournament;
import com.garpr.android.models.FullPlayer;
import com.garpr.android.models.FullTournament;
import com.garpr.android.models.HeadToHead;
import com.garpr.android.models.LiteTournament;
import com.garpr.android.models.Match;
import com.garpr.android.models.MatchesBundle;
import com.garpr.android.models.Ranking;
import com.garpr.android.models.Rating;
import com.garpr.android.models.TournamentsBundle;
import com.garpr.android.models.WinsLosses;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class ListUtils {

    @NonNull
    public static ArrayList<Object> createHeadToHeadList(@NonNull final Context context,
            @Nullable final HeadToHead headToHead) {
        final ArrayList<Object> list = new ArrayList<>();

        if (headToHead == null) {
            list.add(new WinsLosses(0, 0));
            list.add(context.getString(R.string.no_matches));
            list.trimToSize();
            return list;
        }

        list.add(new WinsLosses(headToHead.getWins(), headToHead.getLosses()));

        if (!headToHead.hasMatches()) {
            list.add(context.getString(R.string.no_matches));
            list.trimToSize();
            return list;
        }

        // noinspection ConstantConditions
        list.addAll(createSortedTournamentAndMatchList(headToHead.getMatches()));
        list.trimToSize();

        return list;
    }

    @Nullable
    public static ArrayList<Object> createPlayerMatchesList(@NonNull final Context context,
            @NonNull final RegionManager regionManager, @NonNull final FullPlayer fullPlayer,
            @Nullable final MatchesBundle bundle) {
        final String region = regionManager.getRegion(context);

        // noinspection ConstantConditions
        final Rating rating = fullPlayer.hasRatings() ? fullPlayer.getRatings().getRegion(region)
                : null;

        if (rating == null && (bundle == null || !bundle.hasMatches())) {
            return  null;
        }

        final ArrayList<Object> list = new ArrayList<>();

        if (rating != null) {
            list.add(rating);
        }

        if (bundle == null || !bundle.hasMatches()) {
            if (list.isEmpty()) {
                return null;
            } else {
                list.add(context.getString(R.string.no_matches));
                list.trimToSize();
                return list;
            }
        }

        // noinspection ConstantConditions
        list.addAll(createSortedTournamentAndMatchList(bundle.getMatches()));
        list.trimToSize();

        return list;
    }

    private static ArrayList<Object> createSortedTournamentAndMatchList(
            @NonNull final ArrayList<Match> matches) {
        final ArrayList<Match> matchesCopy = new ArrayList<>(matches);
        Collections.sort(matchesCopy, Match.REVERSE_CHRONOLOGICAL_ORDER);

        final ArrayList<Object> list = new ArrayList<>();
        String tournamentId = null;

        for (final Match match : matchesCopy) {
            if (tournamentId == null || !match.getTournamentId().equals(tournamentId)) {
                tournamentId = match.getTournamentId();
                list.add(new LiteTournament(tournamentId, match.getTournamentName(),
                        match.getTournamentDate()));
            }

            list.add(match);
        }

        return list;
    }

    @Nullable
    public static ArrayList<AbsTournament> createTournamentList(
            @Nullable final TournamentsBundle bundle) {
        if (bundle == null || !bundle.hasTournaments()) {
            return null;
        }

        // noinspection ConstantConditions
        final ArrayList<AbsTournament> tournaments = new ArrayList<>(bundle.getTournaments());
        Collections.sort(tournaments, AbsTournament.REVERSE_CHRONOLOGICAL_ORDER);

        return tournaments;
    }

    @Nullable
    public static ArrayList<Object> filterPlayerMatchesList(@Nullable final Match.Result result,
            @Nullable final List<Object> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        final ArrayList<Object> newList = new ArrayList<>(list.size());

        if (result == null) {
            newList.addAll(list);
            return newList;
        }

        boolean addedCharSequence = false;
        boolean addedRating = false;
        boolean addedWinsLosses = false;

        for (int i = 0; i < list.size(); ++i) {
            final Object objectI = list.get(i);

            if (objectI instanceof CharSequence) {
                if (!addedCharSequence) {
                    addedCharSequence = true;
                    newList.add(objectI);
                }
            } else if (objectI instanceof Rating) {
                if (!addedRating) {
                    addedRating = true;
                    newList.add(objectI);
                }
            } else if (objectI instanceof WinsLosses) {
                if (!addedWinsLosses) {
                    addedWinsLosses = true;
                    newList.add(objectI);
                }
            } else if (objectI instanceof AbsTournament) {
                final AbsTournament tournament = (AbsTournament) objectI;
                boolean addedTournament = false;
                int j = i + 1;

                while (j < list.size()) {
                    final Object objectJ = list.get(j);

                    if (objectJ instanceof Match) {
                        final Match match = (Match) objectJ;

                        if (match.getResult() == result) {
                            if (!addedTournament) {
                                addedTournament = true;
                                newList.add(tournament);
                            }

                            newList.add(match);
                        }

                        ++j;
                    } else {
                        j = list.size();
                    }
                }
            }
        }

        return newList;
    }

    @Nullable
    public static ArrayList<AbsPlayer> searchPlayerList(@Nullable String query,
            @Nullable final List<AbsPlayer> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        final ArrayList<AbsPlayer> newList = new ArrayList<>(list.size());

        if (TextUtils.isEmpty(query) || TextUtils.getTrimmedLength(query) == 0) {
            newList.addAll(list);
            return newList;
        }

        query = query.trim().toLowerCase();

        for (final AbsPlayer player : list) {
            if (player.getName().toLowerCase().contains(query)) {
                newList.add(player);
            }
        }

        return newList;
    }

    @Nullable
    public static ArrayList<Object> searchPlayerMatchesList(@Nullable String query,
            @Nullable final List<Object> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        final ArrayList<Object> newList = new ArrayList<>(list.size());

        if (TextUtils.isEmpty(query) || TextUtils.getTrimmedLength(query) == 0) {
            newList.addAll(list);
            return newList;
        }

        query = query.trim().toLowerCase();
        boolean addedRating = false;

        for (int i = 0; i < list.size(); ++i) {
            final Object objectI = list.get(i);

            if (objectI instanceof Rating) {
                if (!addedRating) {
                    addedRating = true;
                    newList.add(objectI);
                }
            } else if (objectI instanceof AbsTournament) {
                final AbsTournament tournament = (AbsTournament) objectI;
                boolean addedTournament = false;
                int j = i + 1;

                while (j < list.size()) {
                    final Object objectJ = list.get(j);

                    if (objectJ instanceof Match) {
                        final Match match = (Match) objectJ;

                        if (match.getOpponentName().toLowerCase().contains(query)) {
                            if (!addedTournament) {
                                addedTournament = true;
                                newList.add(tournament);
                            }

                            newList.add(match);
                        }

                        ++j;
                    } else {
                        j = list.size();
                    }
                }
            }
        }

        return newList;
    }

    @Nullable
    public static ArrayList<Ranking> searchRankingList(@Nullable String query,
            @Nullable final List<Ranking> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        final ArrayList<Ranking> newList = new ArrayList<>(list.size());

        if (TextUtils.isEmpty(query) || TextUtils.getTrimmedLength(query) == 0) {
            newList.addAll(list);
            return newList;
        }

        query = query.trim().toLowerCase();

        for (final Ranking ranking : list) {
            if (ranking.getName().toLowerCase().contains(query)) {
                newList.add(ranking);
            }
        }

        return newList;
    }

    @Nullable
    public static ArrayList<AbsTournament> searchTournamentList(@Nullable String query,
            @Nullable final List<AbsTournament> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        final ArrayList<AbsTournament> newList = new ArrayList<>(list.size());

        if (TextUtils.isEmpty(query) || TextUtils.getTrimmedLength(query) == 0) {
            newList.addAll(list);
            Collections.sort(newList, AbsTournament.REVERSE_CHRONOLOGICAL_ORDER);
            return newList;
        }

        query = query.trim().toLowerCase();

        for (final AbsTournament tournament : list) {
            if (tournament.getName().toLowerCase().contains(query)) {
                newList.add(tournament);
            }
        }

        Collections.sort(newList, AbsTournament.REVERSE_CHRONOLOGICAL_ORDER);
        return newList;
    }

    @Nullable
    public static ArrayList<FullTournament.Match> searchTournamentMatchesList(@Nullable String query,
            @Nullable final List<FullTournament.Match> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        final ArrayList<FullTournament.Match> newList = new ArrayList<>(list.size());

        if (TextUtils.isEmpty(query) || TextUtils.getTrimmedLength(query) == 0) {
            newList.addAll(list);
            return newList;
        }

        query = query.trim().toLowerCase();

        for (final FullTournament.Match match : list) {
            if (match.getWinnerName().toLowerCase().contains(query) ||
                    match.getLoserName().toLowerCase().contains(query)) {
                newList.add(match);
            }
        }

        return newList;
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/ListUtils.kt
package com.garpr.android.misc

import android.content.Context
import com.garpr.android.R
import com.garpr.android.models.*
import java.util.*

object ListUtils {

    private interface MatchListItemCreator {
        fun createMatch(match: Match): Any
    }

    fun createHeadToHeadList(context: Context, headToHead: HeadToHead?): MutableList<Any>? {
        if (headToHead == null) {
            return null
        }

        val list = mutableListOf<Any>()
        list.add(WinsLosses(headToHead.player, headToHead.wins, headToHead.opponent,
                headToHead.losses))

        val matches = headToHead.matches
        if (matches.isNullOrEmpty()) {
            list.add(context.getString(R.string.these_two_competitors_have_never_played_in_tournament_before))
            return list
        }

        list.addAll(createSortedTournamentAndMatchList(matches, object : MatchListItemCreator {
            override fun createMatch(match: Match): Any {
                return HeadToHeadMatch(match.result, headToHead.player, match.opponent)
            }
        }))

        return list
    }

    fun createPlayerMatchesList(context: Context, fullPlayer: FullPlayer,
            bundle: MatchesBundle?): MutableList<Any>? {
        val newList = mutableListOf<Any>()
        newList.add(fullPlayer)

        val matches = bundle?.matches

        if (matches.isNullOrEmpty()) {
            newList.add(context.getString(R.string.no_matches))
        } else {
            newList.addAll(createSortedTournamentAndMatchList(matches, object : MatchListItemCreator {
                override fun createMatch(match: Match): Any {
                    return match
                }
            }))
        }

        return newList
    }

    fun createRegionsList(bundle: RegionsBundle?): List<Any>? {
        val regions = bundle?.regions

        if (regions.isNullOrEmpty()) {
            return null
        }

        val regionsCopy = mutableListOf<AbsRegion>()
        regionsCopy.addAll(regions)
        Collections.sort(regionsCopy, AbsRegion.ENDPOINT_ORDER)

        val list = mutableListOf<Any>()
        var endpoint: Endpoint? = null

        for (region in regionsCopy) {
            if (region is Region) {
                if (region.isActive) {
                    if (region.endpoint != endpoint) {
                        endpoint = region.endpoint
                        list.add(region.endpoint)
                    }

                    list.add(region)
                }
            } else {
                throw RuntimeException("$region is a ${LiteRegion::class.java.simpleName} " +
                        "when it must be a ${Region::class.java.simpleName}")
            }
        }

        return list
    }

    private fun createSortedTournamentAndMatchList(matches: List<Match>,
            matchListItemCreator: MatchListItemCreator): MutableList<Any> {
        val matchesCopy = mutableListOf<Match>()
        matchesCopy.addAll(matches)
        Collections.sort(matchesCopy, Match.REVERSE_CHRONOLOGICAL_ORDER)

        val list = mutableListOf<Any>()
        var tournamentId: String? = null

        for (match in matchesCopy) {
            if (match.tournament.id != tournamentId) {
                tournamentId = match.tournament.id
                list.add(LiteTournament(null, match.tournament.date, tournamentId,
                        match.tournament.name))
            }

            list.add(matchListItemCreator.createMatch(match))
        }

        return list
    }

    fun createTournamentList(bundle: TournamentsBundle?): List<AbsTournament>? {
        val tournaments = bundle?.tournaments

        return if (tournaments.isNullOrEmpty()) {
            null
        } else {
            val tournamentsCopy = mutableListOf<AbsTournament>()
            tournamentsCopy.addAll(tournaments)
            Collections.sort(tournamentsCopy, AbsTournament.REVERSE_CHRONOLOGICAL_ORDER)
            tournamentsCopy
        }
    }

    fun searchPlayerList(query: String?, list: List<AbsPlayer>?) =
        if (list.isNullOrEmpty()) {
            null
        } else {
            val newList = mutableListOf<AbsPlayer>()

            if (query.isNullOrBlank()) {
                newList.addAll(list)
            } else {
                val trimmedQuery = query.trim()
                newList.addAll(list.filter { it.name.contains(trimmedQuery, true) })
            }

            newList
        }

    fun searchPlayerMatchesList(query: String?, list: List<Any>?): MutableList<Any>? {
        if (list.isNullOrEmpty()) {
            return null
        }

        val newList = mutableListOf<Any>()

        if (query.isNullOrBlank()) {
            newList.addAll(list)
            return newList
        }

        val trimmedQuery = query.trim()
        var addedFullPlayer = false

        for (i in list.indices) {
            val objectI = list[i]

            if (objectI is FullPlayer) {
                if (!addedFullPlayer) {
                    addedFullPlayer = true
                    newList.add(objectI)
                }
            } else if (objectI is AbsTournament) {
                var addedTournament = false
                var j = i + 1

                while (j < list.size) {
                    val objectJ = list[j]

                    if (objectJ is Match) {
                        if (objectJ.opponent.name.contains(trimmedQuery, true)) {
                            if (!addedTournament) {
                                addedTournament = true
                                newList.add(objectI)
                            }

                            newList.add(objectJ)
                        }

                        ++j
                    } else {
                        j = list.size
                    }
                }
            }
        }

        return newList
    }

    fun searchRankingList(query: String?, list: List<RankedPlayer>?) =
        if (list.isNullOrEmpty()) {
            null
        } else {
            val newList = mutableListOf<RankedPlayer>()

            if (query.isNullOrBlank()) {
                newList.addAll(list)
            } else {
                val trimmedQuery = query.trim()
                newList.addAll(list.filter { it.name.contains(trimmedQuery, true) })
            }

            newList
        }

    fun searchTournamentList(query: String?, list: List<AbsTournament>?) =
        if (list.isNullOrEmpty()) {
            null
        } else {
            val newList = mutableListOf<AbsTournament>()

            if (query.isNullOrBlank()) {
                newList.addAll(list)
            } else {
                val trimmedQuery = query.trim()
                newList.addAll(list.filter { it.name.contains(trimmedQuery, true) })
            }

            Collections.sort(newList, AbsTournament.REVERSE_CHRONOLOGICAL_ORDER)
            newList
        }

    fun searchTournamentMatchesList(query: String?, list: List<FullTournament.Match>?) =
        if (list.isNullOrEmpty()) {
            null
        } else {
            val newList = mutableListOf<FullTournament.Match>()

            if (query.isNullOrBlank()) {
                newList.addAll(list)
            } else {
                val trimmedQuery = query.trim()

                newList.addAll(list.filter { it.winnerName.contains(trimmedQuery, true)
                        || it.loserName.contains(trimmedQuery, true) })
            }

            newList
        }

}
